### PR TITLE
Adding GEP-995 for support for named route rules

### LINKED ISSUE #995: GEP: Add Name to HTTPRouteRule and HTTPRouteMatch
**What would you like to be added**:

Add a name field to HTTPRouteRule and HTTPRouteMatch structs to allow users to identify different routes 

**Why this is needed**:

- Policy Attachment

- Route Attachment/Delegation: Ability to have more fine grained matches for policy references IF we add sectionName to the policy object

- Improved Debugging: Ability to add debugging information, such as adding the route matched to some access log

### PR DESCRIPTION
**Pull request by guicassolato**
**What type of PR is this?**
/kind gep
/kind api-change

**What this PR does / why we need it**:
Proposal to add a new `name` field to the route rule types ([GRPCRouteRule](https://gateway-api.sigs.k8s.io/reference/spec/#gateway.networking.k8s.io/v1alpha2.GRPCRouteRule), [HTTPRouteRule](https://gateway-api.sigs.k8s.io/reference/spec/#gateway.networking.k8s.io/v1.HTTPRouteRule), [TCPRouteRule](https://gateway-api.sigs.k8s.io/reference/spec/#gateway.networking.k8s.io/v1alpha2.TCPRouteRule), [TLSRouteRule](https://gateway-api.sigs.k8s.io/reference/spec/#gateway.networking.k8s.io/v1alpha2.TLSRouteRule) and [UDPRouteRule](https://gateway-api.sigs.k8s.io/reference/spec/#gateway.networking.k8s.io/v1alpha2.UDPRouteRule)) to support referencing individual rules by name.

**Which issue(s) this PR fixes**:
Part of #995

**Does this PR introduce a user-facing change?**:
```release-note
Add a new `name` field to the route rule types GRPCRouteRule, HTTPRouteRule, TCPRouteRule, TLSRouteRule, and UDPRouteRule.
```

### CODE REVIEW

**Comment by guicassolato** [comment]
Can we please get some preliminary feedback on this (Goals/Non-goals, especially), so we can try moving on with the actual proposal?

Maybe @youngnick and/or @robscott please? 🙏 (If you have the time for it. I know you're usually very busy.)

**Comment by kflynn** [comment]
@guicassolato, thanks for putting this together! I'm going to try to pay more attention to it, as we're finding a use case for this in Linkerd, too. 

One questions for you that's not in the draft GEP  -- are you thinking that must names be unique?

**Comment by spacewander** [comment]
Is there any update on this PR?
IMHO, many people like us need the new `name` filed in the xRoute.

**Comment by guicassolato** [comment]
> Is there any update on this PR?

There was no objection regarding the goals/non-goals for a couple of weeks, so I've pushed the proposal for the API changes. Now, it needs review.

**Comment by spacewander** [comment]
@kflynn @robscott @howardjohn @hzxuzhonghu 
This GEP targets https://github.com/kubernetes-sigs/gateway-api/issues/995. Would you take a look at it? Thanks!

File: geps/gep-995.md
Comment (mikemorris) [inline geps/gep-995.md:34]:
```suggestion
Referencing list elements without a unique identifier is also prone to execution errors, either when relying on how the elements are sorted in the list (i.e., based on the index) or on partial or total repetition of values of the referents. The order of elements within a list may change without necessarily any semantic reason. Complex elements can sometimes differ only subtly from each other, thus easily being overlooked when making the reference and resulting in a higher chance of typos and/or references that are possibly ambiguous or broken. In both cases, such references are fragile and can result in unexpected errors.
```

I think the notes about "repetition" here are a bit too abstract to be easily understood, and those sentences could probably be dropped without losing any important detail - this is covered more succictly and clearly in the Alternatives section. The relevant bits about index reference inconsistency are already covered well in the last paragraph. Removed list formatting.

File: geps/gep-995.md
Comment (mikemorris) [inline geps/gep-995.md:36]:
```suggestion
For the `Gateway` resource, problems above were addressed/mitigated by adding a `name` field to the [`Listener`](https://gateway-api.sigs.k8s.io/reference/spec/#gateway.networking.k8s.io/v1.Listener) type ([#724](https://github.com/kubernetes-sigs/gateway-api/issues/).) Listener names are required and must be unique of each listener declared in a gateway. This allowed for more explicit route and policy attachment relying on _sectionName_, as well as it opened for better implementation of status reporting and log recording of events related to specific gateway listeners.
```

File: geps/gep-995.md
Comment (mikemorris) [inline geps/gep-995.md:40]:
```suggestion
This GEP aims to rollout the same pattern of declarative `name` fields of these examples to the Gateway API route rule types.
```
Just feels slightly redundant as these are repeated in the next section.

File: geps/gep-995.md
Comment (mikemorris) [inline geps/gep-995.md:44]:
```suggestion
This GEP proposes to add a new optional `name` field to the [GRPCRouteRule](https://gateway-api.sigs.k8s.io/reference/spec/#gateway.networking.k8s.io/v1alpha2.GRPCRouteRule), [HTTPRouteRule](https://gateway-api.sigs.k8s.io/reference/spec/#gateway.networking.k8s.io/v1.HTTPRouteRule), [TCPRouteRule](https://gateway-api.sigs.k8s.io/reference/spec/#gateway.networking.k8s.io/v1alpha2.TCPRouteRule), [TLSRouteRule](https://gateway-api.sigs.k8s.io/reference/spec/#gateway.networking.k8s.io/v1alpha2.TLSRouteRule) and [UDPRouteRule](https://gateway-api.sigs.k8s.io/reference/spec/#gateway.networking.k8s.io/v1alpha2.UDPRouteRule) types.
```

I think we should probably just do this on all route types to be consistent.

File: geps/gep-995.md
Comment (mikemorris) [inline geps/gep-995.md:66]:
```suggestion
Implementations MAY recomend the usage of the `name` field for enabling specific features, such as for supporting policy attachment targetting individual route rules, and more assertive log messages and/or status reporting that include on the name of the rule. However, because as by API design the presence of the field is optional, implementations MUST take into account that a value may sometimes not be available. For such cases, implementations are free to decide whether to provide the feature depending the `name` field, if the feature is not required for Core compliance, or to enable the feature relying on another method of referencing of choice.
```

File: geps/gep-995.md
Comment (mikemorris) [inline geps/gep-995.md:70]:
```suggestion
Implementations MUST NOT assume a default value for the `name` field for route rules when not specified by the end-user.
```

File: geps/gep-995.md
Comment (mikemorris) [inline geps/gep-995.md:78]: This looks like the last section missing before moving this out of draft?

File: geps/gep-995.md
Comment (mikemorris) [inline geps/gep-995.md:98]: Probably worth citing https://github.com/kubernetes-sigs/gateway-api/issues/1696#issuecomment-1679804122 here too - even though large routes with many rules should be discouraged, this is still useful with small routes with just a few rules.

**Review by mikemorris — APPROVED** [review]
Overall LGTM, thanks for writing this up @guicassolato! Minor notes, biggest change being to explicit apply this change to _all_ route types. I hope we can discuss this at the first Gateway API meeting in 2024 and get this merged soon!

**Comment by guicassolato** [comment]
Thanks for the suggestions @mikemorris! I'm officially back from holidays tomorrow and I'll make the changes to the PR.

File: geps/gep-995.md
Comment (guicassolato) [reply geps/gep-995.md]: Just added a short bullet point to one of the alternatives, to cover the part of the meaning that I believe could otherwise be lost, and accepted the suggestion. Thanks!

File: geps/gep-995/metadata.yaml
Comment (guicassolato) [inline geps/gep-995/metadata.yaml:8]: @youngnick, add relationships to Policy Attachment here, maybe once finishing splitting GEP-713 into [Direct](https://github.com/kubernetes-sigs/gateway-api/issues/2648) and [Inherited](https://github.com/kubernetes-sigs/gateway-api/issues/2649)?

Comment (guicassolato) [reply geps/gep-995/metadata.yaml:8]: Done it.

**Comment by robscott** [comment]
Sorry we've taken so long to get to this @guicassolato! I think this has been very well written, and given our recent discussions at KubeCon about this, I think we're overdue to get this in. With that said, I think we're realistically a bit too late to hit the v1.1 cut, but hopefully this can be one of the first features included in v1.2.

/lgtm
/approve

**Comment by robscott** [comment]
Since we've got approvals from a couple maintainers on this I think we're good to remove the hold.

/hold cancel

**Pull request merged**
